src/features/semanticTokensProvider.ts

```diff
@@ -0,0 +1,315 @@
+    namespace,
+    type,
+    struct,
+    class,
+    interface,
+    enum,
+    typeParameter,
+    function,
+    member,
+    macro,
+    variable,
+    parameter,
+    property,
+    enumMember,
+    event,
+    label,
+}
+
+enum CustomTokenTypes {
+    plainKeyword = DefaultTokenType.label + 1,
+    controlKeyword,
+    operatorOverloaded,
+    preprocessorKeyword,
+    preprocessorText,
+    punctuation,
+    stringVerbatim,
+    stringEscapeCharacter,
+    delegate,
+    module,
+    extensionMethod,
+    field,
+    local
+}
+
+// The default TokenModifiers defined by VS Code https://github.com/microsoft/vscode/blob/master/src/vs/platform/theme/common/tokenClassificationRegistry.ts#L393
+enum DefaultTokenModifier {
+    declaration,
+    static,
+    abstract,
+    deprecated,
+    modification,
+    async,
+    readonly,
+}
+
+// All classifications from Roslyn's ClassificationTypeNames https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs
+// Keep in sync with omnisharp-roslyn's SemanticHighlightClassification
+enum SemanticHighlightClassification {
+    Comment,
+    ExcludedCode,
+    Identifier,
+    Keyword,
+    ControlKeyword,
+    NumericLiteral,
+    Operator,
+    OperatorOverloaded,
+    PreprocessorKeyword,
+    StringLiteral,
+    WhiteSpace,
+    Text,
+    StaticSymbol,
+    PreprocessorText,
+    Punctuation,
+    VerbatimStringLiteral,
+    StringEscapeCharacter,
+    ClassName,
+    DelegateName,
+    EnumName,
+    InterfaceName,
+    ModuleName,
+    StructName,
+    TypeParameterName,
+    FieldName,
+    EnumMemberName,
+    ConstantName,
+    LocalName,
+    ParameterName,
+    MethodName,
+    ExtensionMethodName,
+    PropertyName,
+    EventName,
+    NamespaceName,
+    LabelName,
+    XmlDocCommentAttributeName,
+    XmlDocCommentAttributeQuotes,
+    XmlDocCommentAttributeValue,
+    XmlDocCommentCDataSection,
+    XmlDocCommentComment,
+    XmlDocCommentDelimiter,
+    XmlDocCommentEntityReference,
+    XmlDocCommentName,
+    XmlDocCommentProcessingInstruction,
+    XmlDocCommentText,
+    XmlLiteralAttributeName,
+    XmlLiteralAttributeQuotes,
+    XmlLiteralAttributeValue,
+    XmlLiteralCDataSection,
+    XmlLiteralComment,
+    XmlLiteralDelimiter,
+    XmlLiteralEmbeddedExpression,
+    XmlLiteralEntityReference,
+    XmlLiteralName,
+    XmlLiteralProcessingInstruction,
+    XmlLiteralText,
+    RegexComment,
+    RegexCharacterClass,
+    RegexAnchor,
+    RegexQuantifier,
+    RegexGrouping,
+    RegexAlternation,
+    RegexText,
+    RegexSelfEscapedCharacter,
+    RegexOtherEscape,
+}
+
+enum SemanticHighlightModifier {
+    Static
+}
+
+export default class SemanticTokensProvider extends AbstractProvider implements vscode.DocumentSemanticTokensProvider, vscode.DocumentRangeSemanticTokensProvider {
+
+    getLegend(): vscode.SemanticTokensLegend {
+        return new vscode.SemanticTokensLegend(tokenTypes, tokenModifiers);
```

Comment:
The server's typically the one that defines the legend. Why re-create it on the client (vs. querying it)?

Reply:
Yeah goes back to whether the OmniSharp endpoint should just implement the LSP return types. I'll think on it.